src/Text/Parsing/Parser.purs

```diff
@@ -13,74 +13,77 @@ import Data.Tuple (Tuple(..))
 import Text.Parsing.Parser.Pos (Position, initialPos)
 
 -- | A parsing error, consisting of a message and position information.
-data ParseError = ParseError
-  { message :: String
-  , position :: Position
-  }
+data ParseError = ParseError String Position
+
+-- | The result of a single parse
+data Result s a = Result  s                     -- the new input
+                          (Either ParseError a) -- the result
+                          Boolean               -- consumed?
+                          Position              -- the new position
 
 instance showParseError :: Show ParseError where
-  show (ParseError msg) = "ParseError { message: " <> msg.message <> ", position: " <> show msg.position <> " }"
+  show (ParseError msg pos) = "ParseError " <> msg <> " " <> show pos
 
 instance eqParseError :: Eq ParseError where
-  eq (ParseError {message : m1, position : p1}) (ParseError {message : m2, position : p2}) = m1 == m2 && p1 == p2
+  eq (ParseError m1  p1) (ParseError m2 p2) = m1 == m2 && p1 == p2
 
 -- | `PState` contains the remaining input and current position.
-data PState s = PState
-  { input :: s
-  , position :: Position
-  }
+data PState s = PState s Position
 
 -- | The Parser monad transformer.
 -- |
 -- | The first type argument is the stream type. Typically, this is either `String`, or some sort of token stream.
-newtype ParserT s m a = ParserT (PState s -> m { input :: s, result :: Either ParseError a, consumed :: Boolean, position :: Position })
+newtype ParserT s m a = ParserT (PState s -> m (Result s a))
 
 -- | Apply a parser by providing an initial state.
-unParserT :: forall m s a. ParserT s m a -> PState s -> m { input :: s, result :: Either ParseError a, consumed :: Boolean, position :: Position }
+unParserT :: forall m s a. ParserT s m a -> PState s -> m (Result s a)
 unParserT (ParserT p) = p
 
 -- | Apply a parser, keeping only the parsed result.
 runParserT :: forall m s a. Monad m => PState s -> ParserT s m a -> m (Either ParseError a)
 runParserT s p = do
-  o <- unParserT p s
-  pure o.result
+  (Result _ result _ _) <- unParserT p s
+  pure result
 
 -- | The `Parser` monad is a synonym for the parser monad transformer applied to the `Identity` monad.
 type Parser s a = ParserT s Identity a
 
 -- | Apply a parser, keeping only the parsed result.
 runParser :: forall s a. s -> Parser s a -> Either ParseError a
-runParser s = runIdentity <<< runParserT (PState { input: s, position: initialPos })
+runParser s = runIdentity <<< runParserT (PState s initialPos)
```

Comment:
The expression `PState s initialPos` is what should be pulled out into a smart constructor for `PState`.